### DISCUSSION TITLE
pyransame - Python RAndom SAmpling of MEshes using PyVista

### DISCUSSION THREAD

**Discussion by MatthewFlamm** [pyvista/pyvista#4292] [Show and tell]
I have use cases where I need to sample random points inside cells of a mesh.  One example is to post-process a Eulerian description of particle motion, e.g. with a volume fraction, to be shown with representative particles located with an appropriate spatial density.  Another example is to provide random starting locations on a surface for streamline generation.  Sometimes a random weighting provides a pleasing streamline plot with some "natural" variability as opposed to structured starting locations.  Sometimes I also need to generate streamlines to capture a particular feature of a flow, and random starting locations provide a mechanism to easily iterate through many starting points to find the most relevant streamline.

I have started [pyransame](https://matthewflamm.github.io/pyransame/) to fill a gap I have noticed in vtk, where this functionality is not easily accomplished.  Currently, it only works on `PolyData` meshes which can be triangulated, but I will be looking into 3D meshes next.

See also the [github repo](https://github.com/MatthewFlamm/pyransame).
![image](https://user-images.githubusercontent.com/39341281/232538712-01e51ff8-ab31-45a7-934b-44e0ed078930.png)

**Comment by banesullivan** [comment]
This is nice -- indeed a great gap to fill in VTK! Thanks for sharing!

**Comment by MatthewFlamm** [comment]
pyransame now supports random sampling in 3D volumes.  It currently supports a handful of cell types for meshes (both 2D and 3D) and more to added in the future.

![image](https://user-images.githubusercontent.com/39341281/234640993-fc5cdea3-fdfc-4471-bd6c-216b9cc86965.png)

**Reply by MatthewFlamm** [reply]
Version 0.3.0 now supports automatic sampling of point and cell data with the improvements in PyVista to the `sample` filter on the last two releases.  1D cells are also now supported.

**Reply by MatthewFlamm** [reply]
Nearly all linear 3D, 2D, and 1D cells are now supported as well

**Reply by MatthewFlamm** [reply]
v0.4.0 recently published that require features from pyvista 0.44. All linear cells (except for one) including 0D cells are now supported.

**Comment by tkoyama010** [comment]
Awesome job! This sounds useful for CFD person. I would introduce it to my friends in that field.
